File: catalog.go
package main

import (
	"bufio"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

type catalogDetection struct {
	Packages []string `json:"packages,omitempty"`
	Services []string `json:"services,omitempty"`
	Paths    []string `json:"paths,omitempty"`
}

type catalogCompatibility struct {
	Status string   `json:"status"`
	Hints  []string `json:"hints,omitempty"`
}

type catalogInstallPlan struct {
	Method        string   `json:"method,omitempty"`
	Repository    string   `json:"repository,omitempty"`
	RepositoryURL string   `json:"repository_url,omitempty"`
	Packages      []string `json:"packages,omitempty"`
	InstallerURL  string   `json:"installer_url,omitempty"`
	Notes         []string `json:"notes,omitempty"`
	PreviewOnly   bool     `json:"preview_only"`
}

type catalogItem struct {
	ID             string               `json:"id"`
	Kind           string               `json:"kind"`
	Name           string               `json:"name"`
	Category       string               `json:"category"`
	Description    string               `json:"description"`
	ProjectURL     string               `json:"project_url,omitempty"`
	Source         string               `json:"source"`
	State          string               `json:"state"`
	Installed      bool                 `json:"installed"`
	Enabled        bool                 `json:"enabled"`
	Version        string               `json:"version,omitempty"`
	Service        string               `json:"service,omitempty"`
	ServiceRunning bool                 `json:"service_running"`
	WebPort        int                  `json:"web_port,omitempty"`
	WebPortSource  string               `json:"web_port_source,omitempty"`
	Capabilities   []string             `json:"capabilities,omitempty"`
	Detection      catalogDetection     `json:"detection,omitempty"`
	Compatibility  catalogCompatibility `json:"compatibility"`
	Install        catalogInstallPlan   `json:"install,omitempty"`

	ProcessNames       []string `json:"-"`
	WebRequiresPackage string   `json:"-"`
	Builtin            bool     `json:"-"`
	Planned            bool     `json:"-"`
}

type catalogSnapshot struct {
	GeneratedAt  time.Time     `json:"generated_at"`
	ReadOnly     bool          `json:"read_only"`
	Phase        string        `json:"phase"`
	Modules      []catalogItem `json:"modules"`
	Integrations []catalogItem `json:"integrations"`
}

func readCatalog() catalogSnapshot {
	return buildCatalog(readInstalledPackages(), readProcessNames(), pathExists)
}

func buildCatalog(installed map[string]string, processes map[string]bool, exists func(string) bool) catalogSnapshot {
	modules := builtinModuleCatalog()
	integrations := integrationCatalog()

	for i := range modules {
		finalizeCatalogItem(&modules[i], installed, processes, exists)
	}
	for i := range integrations {
		finalizeCatalogItem(&integrations[i], installed, processes, exists)
	}

	sort.Slice(modules, func(i, j int) bool {
		if modules[i].Category != modules[j].Category {
			return modules[i].Category < modules[j].Category
		}
		return modules[i].Name < modules[j].Name
	})
	sort.Slice(integrations, func(i, j int) bool {
		if integrations[i].Category != integrations[j].Category {
			return integrations[i].Category < integrations[j].Category
		}
		return integrations[i].Name < integrations[j].Name
	})

	return catalogSnapshot{
		GeneratedAt:  time.Now(),
		ReadOnly:     true,
		Phase:        "catalog-foundation",
		Modules:      modules,
		Integrations: integrations,
	}
}

func builtinModuleCatalog() []catalogItem {
	return []catalogItem{
		{
			ID:            "dns-core",
			Kind:          "module",
			Name:          "DNS Core",
			Category:      "Core",
			Description:   "Базовый DNS-мониторинг, корреляция клиентов, resolver health и диагностика маршрутизации.",
			Source:        "builtin",
			Builtin:       true,
			Enabled:       true,
			Capabilities:  []string{"dns-observability", "client-attribution", "routing-diagnostics"},
			Compatibility: catalogCompatibility{Status: "built-in"},
		},
		{
			ID:            "marketplace",
			Kind:          "module",
			Name:          "Marketplace",
			Category:      "Platform",
			Description:   "Каталог модулей и сторонних интеграций. На первом этапе работает только в режиме обнаружения и предпросмотра.",
			Source:        "builtin",
			Builtin:       true,
			Enabled:       true,
			Capabilities:  []string{"catalog", "integration-detection", "install-plan-preview"},
			Compatibility: catalogCompatibility{Status: "built-in"},
		},
		plannedModule("system", "System Monitor", "Monitoring", "CPU по ядрам, RAM, uptime, load average и системная сводка."),
		plannedModule("thermal", "Thermal Monitor", "Monitoring", "Температуры по всем доступным thermal/hwmon датчикам."),
		plannedModule("storage", "Storage Monitor", "Monitoring", "Файловые системы, свободное место, I/O, USB и состояние хранилищ."),
		plannedModule("network", "Network Monitor", "Monitoring", "Интерфейсы, RX/TX, ошибки и сетевые счётчики."),
		plannedModule("admin", "Admin Tools", "Administration", "Процессы, службы, opkg, файлы и терминал. Потребует отдельной авторизации."),
		plannedModule("profiling", "Profiling", "Development", "pprof, slow-request logging и внутренняя диагностика DNS Monitor."),
	}
}

func plannedModule(id, name, category, description string) catalogItem {
	return catalogItem{
		ID:            id,
		Kind:          "module",
		Name:          name,
		Category:      category,
		Description:   description,
		Source:        "dns-monitor",
		Planned:       true,
		Capabilities:  []string{"planned"},
		Compatibility: catalogCompatibility{Status: "planned"},
	}
}

func integrationCatalog() []catalogItem {
	return []catalogItem{
		{
			ID:           "awg-manager",
			Kind:         "integration",
			Name:         "AWG Manager",
			Category:     "VPN / Routing",
			Description:  "Управление AmneziaWG и sing-box туннелями на Keenetic.",
			ProjectURL:   "https://github.com/hoaxisr/awg-manager",
			Source:       "official",
			Capabilities: []string{"detect", "version", "service-status", "open-ui", "install-preview"},
			Detection: catalogDetection{
				Packages: []string{"awg-manager"},
				Services: []string{"/opt/etc/init.d/S99awg-manager"},
				Paths:    []string{"/opt/etc/awg-manager"},
			},
			ProcessNames:  []string{"awg-manager"},
			WebPort:       2222,
			WebPortSource: "default-hint",
			Compatibility: catalogCompatibility{
				Status: "requirements",
				Hints:  []string{"Keenetic / Netcraze", "Entware", "Компонент WireGuard KeeneticOS"},
			},
			Install: catalogInstallPlan{
				Method:       "official-script",
				InstallerURL: "http://repo.hoaxisr.ru/install.sh",
				Packages:     []string{"awg-manager"},
				Notes:        []string{"Официальный install.sh AWG Manager.", "Выполнение из DNS Monitor пока отключено."},
				PreviewOnly:  true,
			},
		},
		{
			ID:           "nfqws",
			Kind:         "integration",
			Name:         "nfqws",
			Category:     "DPI / Bypass",
			Description:  "Классическая nfqws-keenetic интеграция. Поддерживается обнаружение; для новых установок предпочтительнее nfqws2.",
			ProjectURL:   "https://github.com/Anonym-tsk/nfqws-keenetic",
			Source:       "official",
			Capabilities: []string{"detect", "version", "service-status", "open-ui"},
			Detection: catalogDetection{
				Packages: []string{"nfqws-keenetic"},
				Services: []string{"/opt/etc/init.d/S51nfqws", "/opt/etc/init.d/S99zapret"},
				Paths:    []string{"/opt/etc/nfqws"},
			},
			ProcessNames:       []string{"nfqws"},
			WebPort:            90,
			WebPortSource:      "nfqws-keenetic-web",
			WebRequiresPackage: "nfqws-keenetic-web",
			Compatibility: catalogCompatibility{
				Status: "requirements",
				Hints:  []string{"Entware", "Netfilter kernel modules", "Legacy: nfqws2 является новой веткой"},
			},
			Install: catalogInstallPlan{
				Method:      "manual",
				Notes:       []string{"На первом этапе Marketplace только обнаруживает legacy nfqws.", "Автоматическую установку добавим после отдельной проверки официального feed."},
				PreviewOnly: true,
			},
		},
		{
			ID:           "nfqws2",
			Kind:         "integration",
			Name:         "nfqws2",
			Category:     "DPI / Bypass",
			Description:  "Новая версия nfqws для NFQUEUE/raw sockets с Entware-пакетом и отдельным web UI.",
			ProjectURL:   "https://github.com/nfqws/nfqws2-keenetic",
			Source:       "official",
			Capabilities: []string{"detect", "version", "service-status", "open-ui", "install-preview", "conflict-detection"},
			Detection: catalogDetection{
				Packages: []string{"nfqws2-keenetic"},
				Services: []string{"/opt/etc/init.d/S51nfqws2"},
				Paths:    []string{"/opt/etc/nfqws2"},
			},
			ProcessNames:       []string{"nfqws2"},
			WebPort:            90,
			WebPortSource:      "nfqws-keenetic-web",
			WebRequiresPackage: "nfqws-keenetic-web",
			Compatibility: catalogCompatibility{
				Status: "requirements",
				Hints:  []string{"Entware", "Netfilter kernel modules", "Конфликтует с legacy nfqws-keenetic"},
			},
			Install: catalogInstallPlan{
				Method:        "opkg-feed",
				Repository:    "nfqws2-keenetic",
				RepositoryURL: "https://nfqws.github.io/nfqws2-keenetic/all",
				Packages:      []string{"nfqws2-keenetic"},
				Notes:         []string{"Перед миграцией legacy nfqws должен быть удалён.", "Web UI nfqws-keenetic-web устанавливается отдельно."},
				PreviewOnly:   true,
			},
		},
		{
			ID:           "hydraroute-neo",
			Kind:         "integration",
			Name:         "HydraRoute Neo",
			Category:     "Routing",
			Description:  "Маршрутизация по доменам/CIDR и L7 (TLS SNI / HTTP Host / QUIC) через политики или интерфейсы Keenetic.",
			ProjectURL:   "https://github.com/Ground-Zerro/HydraRoute",
			Source:       "official",
			Capabilities: []string{"detect", "version", "service-status", "open-ui", "install-preview"},
			Detection: catalogDetection{
				Packages: []string{"hrneo"},
				Services: []string{"/opt/etc/init.d/S99hrneo"},
				Paths:    []string{"/opt/etc/HydraRoute"},
			},
			ProcessNames:  []string{"hrneo"},
			WebPort:       2000,
			WebPortSource: "default",
			Compatibility: catalogCompatibility{
				Status: "requirements",
				Hints:  []string{"KeeneticOS > 4.3.6", "Entware", "Xtables-addons для Netfilter"},
			},
			Install: catalogInstallPlan{
				Method:       "official-script",
				InstallerURL: "https://git.zerrolabs.org/Ground-Zerro/release/pages/keenetic/install-neo.sh",
				Packages:     []string{"hrneo"},
				Notes:        []string{"Официальный installer HydraRoute Neo также ставит HRweb.", "Выполнение из DNS Monitor пока отключено."},
				PreviewOnly:  true,
			},
		},
	}
}

func finalizeCatalogItem(item *catalogItem, installed map[string]string, processes map[string]bool, exists func(string) bool) {
	if item.Builtin {
		item.Installed = true
		item.State = "installed"
		return
	}
	if item.Planned {
		item.State = "planned"
		return
	}

	for _, pkg := range item.Detection.Packages {
		if version, ok := installed[pkg]; ok {
			item.Installed = true
			if item.Version == "" {
				item.Version = version
			}
		}
	}
	for _, service := range item.Detection.Services {
		if exists(service) {
			item.Installed = true
			if item.Service == "" {
				item.Service = service
			}
		}
	}
	for _, p := range item.Detection.Paths {
		if exists(p) {
			item.Installed = true
		}
	}
	for _, name := range item.ProcessNames {
		if processes[strings.ToLower(name)] {
			item.ServiceRunning = true
			break
		}
	}

	if item.WebRequiresPackage != "" {
		if _, ok := installed[item.WebRequiresPackage]; !ok {
			item.WebPort = 0
		}
	}

	if item.Installed {
		item.State = "installed_external"
		item.Enabled = item.ServiceRunning
	} else {
		item.State = "available"
	}
}

func readInstalledPackages() map[string]string {
	paths := []string{
		"/opt/lib/opkg/status",
		"/opt/var/opkg/status",
	}

	for i := range paths {
		p := paths[i]
		f, err := os.Open(p)
		if err != nil {
			continue
		}

		out := parseOpkgStatus(f)
		f.Close()

		if len(out) > 0 {
			return out
		}
	}

	return map[string]string{}
}

func parseOpkgStatus(r io.Reader) map[string]string {
	out := map[string]string{}
	sc := bufio.NewScanner(r)
	var pkg, version string
	commit := func() {
		if pkg != "" {
			out[pkg] = version
		}
		pkg, version = "", ""
	}
	for sc.Scan() {
		line := sc.Text()
		if strings.TrimSpace(line) == "" {
			commit()
			continue
		}
		if strings.HasPrefix(line, "Package:") {
			pkg = strings.TrimSpace(strings.TrimPrefix(line, "Package:"))
		}
		if strings.HasPrefix(line, "Version:") {
			version = strings.TrimSpace(strings.TrimPrefix(line, "Version:"))
		}
	}
	commit()
	return out
}

func readProcessNames() map[string]bool {
	out := map[string]bool{}
	entries, err := os.ReadDir("/proc")
	if err != nil {
		return out
	}
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		if _, err := strconv.Atoi(entry.Name()); err != nil {
			continue
		}
		base := filepath.Join("/proc", entry.Name())
		if b, err := os.ReadFile(filepath.Join(base, "comm")); err == nil {
			name := strings.ToLower(strings.TrimSpace(string(b)))
			if name != "" {
				out[name] = true
			}
		}
		if b, err := os.ReadFile(filepath.Join(base, "cmdline")); err == nil {
			parts := strings.Split(string(b), "\x00")
			if len(parts) > 0 && parts[0] != "" {
				name := strings.ToLower(filepath.Base(parts[0]))
				if name != "" {
					out[name] = true
				}
			}
		}
	}
	return out
}

func pathExists(p string) bool {
	_, err := os.Stat(p)
	return err == nil
}

File: web/js/pages/catalog.js
import { esc } from '../utils.js';

function stateInfo(item){
  switch(item.state){
    case 'installed': return {label:'Установлен', cls:'ok'};
    case 'installed_external': return {label:item.service_running?'Установлен · работает':'Установлен', cls:item.service_running?'ok':'warn'};
    case 'planned': return {label:'Запланирован', cls:'accent'};
    case 'incompatible': return {label:'Несовместим', cls:'error'};
    default: return {label:'Доступен', cls:''};
  }
}

function localWebURL(port){
  if(!port)return '';
  const host=location.hostname.includes(':')?`[${location.hostname}]`:location.hostname;
  return `http://${host}:${port}`;
}

function capabilityLabel(v){
  const map={
    detect:'обнаружение', version:'версия', 'service-status':'служба',
    'open-ui':'web UI', 'install-preview':'план установки',
    'conflict-detection':'конфликты', catalog:'каталог',
    'integration-detection':'интеграции', 'install-plan-preview':'план установки',
    planned:'в разработке'
  };
  return map[v]||v;
}

function installPlan(item){
  const p=item.install||{};
  if(!p.method)return '';
  const rows=[];
  if(p.repository_url)rows.push(`<div><span>Feed</span><code>${esc(p.repository_url)}</code></div>`);
  if(p.installer_url)rows.push(`<div><span>Installer</span><code>${esc(p.installer_url)}</code></div>`);
  if((p.packages||[]).length)rows.push(`<div><span>Пакеты</span><code>${esc(p.packages.join(', '))}</code></div>`);
  const notes=(p.notes||[]).map(n=>`<li>${esc(n)}</li>`).join('');
  return `<details class="catalog-plan"><summary>План установки</summary>
    <div class="catalog-plan-grid"><div><span>Метод</span><code>${esc(p.method)}</code></div>${rows.join('')}</div>
    ${notes?`<ul>${notes}</ul>`:''}
    <div class="catalog-readonly-note">Предпросмотр: выполнение install/update/remove на этом этапе отключено.</div>
  </details>`;
}

function card(item){
  const st=stateInfo(item);
  const caps=(item.capabilities||[]).map(c=>`<span class="pill">${esc(capabilityLabel(c))}</span>`).join('');
  const hints=(item.compatibility?.hints||[]).map(h=>`<li>${esc(h)}</li>`).join('');
  const web=item.installed&&item.web_port
    ?`<a class="btn" target="_blank" rel="noopener" href="${esc(localWebURL(item.web_port))}">Открыть UI :${item.web_port}</a>`
    :'';
  const version=item.version?`<span class="catalog-version">v${esc(item.version)}</span>`:'';
  const source=item.source==='official'?'официальный проект':item.source==='builtin'?'встроено':'DNS Monitor';
  return `<article class="catalog-card">
    <div class="catalog-card-head">
      <div><div class="catalog-card-title">${esc(item.name)} ${version}</div><div class="catalog-card-category">${esc(item.category)} · ${esc(source)}</div></div>
      <span class="pill ${st.cls}">${st.label}</span>
    </div>
    <div class="catalog-card-description">${esc(item.description||'')}</div>
    ${item.service?`<div class="catalog-service mono">${esc(item.service)}</div>`:''}
    <div class="catalog-capabilities">${caps}</div>
    ${hints?`<div class="catalog-requirements"><div>Совместимость / требования</div><ul>${hints}</ul></div>`:''}
    ${installPlan(item)}
    <div class="catalog-card-actions">
      ${web}
      ${item.project_url?`<a class="btn" target="_blank" rel="noopener" href="${esc(item.project_url)}">Проект</a>`:''}
    </div>
  </article>`;
}

export function renderCatalog(data,ui){
  const all=[...(data.modules||[]),...(data.integrations||[])];
  const categories=[...new Set(all.map(x=>x.category).filter(Boolean))].sort((a,b)=>a.localeCompare(b,'ru'));
  const q=(ui.catalogSearch||'').trim().toLowerCase();
  const kind=ui.catalogKind||'all';
  const category=ui.catalogCategory||'all';
  const items=all.filter(x=>{
    if(kind!=='all'&&x.kind!==kind)return false;
    if(category!=='all'&&x.category!==category)return false;
    if(!q)return true;
    return `${x.name} ${x.category} ${x.description} ${(x.capabilities||[]).join(' ')}`.toLowerCase().includes(q);
  });
  const categoryOptions=`<option value="all">Все категории</option>${categories.map(c=>`<option value="${esc(c)}" ${category===c?'selected':''}>${esc(c)}</option>`).join('')}`;
  return `<h1 class="page-title">Модули и интеграции</h1>
    <div class="catalog-banner">
      <strong>Marketplace foundation</strong>
      <span>Сейчас каталог работает только на чтение: обнаруживает установленные системы и показывает безопасный план будущей установки. Никаких opkg/install/remove из web пока не выполняется.</span>
    </div>
    <div class="toolbar">
      <input class="input" id="catalogSearch" placeholder="Поиск модуля или интеграции..." value="${esc(ui.catalogSearch||'')}">
      <select class="select" id="catalogKind">
        <option value="all" ${kind==='all'?'selected':''}>Всё</option>
        <option value="module" ${kind==='module'?'selected':''}>Модули DNS Monitor</option>
        <option value="integration" ${kind==='integration'?'selected':''}>Сторонние интеграции</option>
      </select>
      <select class="select" id="catalogCategory">${categoryOptions}</select>
    </div>
    <div class="catalog-summary">
      <span>${items.length} элементов</span>
      <span>${all.filter(x=>x.state==='installed_external'||x.state==='installed').length} обнаружено</span>
      <span>режим: read-only</span>
    </div>
    <div class="catalog-grid">${items.map(card).join('')||'<div class="panel"><div class="empty">Ничего не найдено</div></div>'}</div>`;
}
